File: src/types/organization.ts
/**
 * Multi-tenant Organization Model
 *
 * Architecture overview:
 * - Organization (tenant) is the top-level entity
 * - Users belong to an organization with a specific role
 * - Courses are scoped to an organization
 * - Admins assign courses to employees or make them org-wide
 */

export type OrgRole = 'owner' | 'admin' | 'instructor' | 'learner'

export type OrgPlan = 'starter' | 'professional' | 'enterprise'

export interface Organization {
  id: string
  name: string
  slug: string // URL-friendly identifier, e.g. "acme-corp"
  logo?: string
  domain?: string // e.g. "acme.com" — for auto-joining via email domain
  plan: OrgPlan
  settings: OrgSettings
  createdAt: Date
}

export interface OrgSettings {
  maxUsers: number
  maxCourses: number
  allowSelfEnrollment: boolean // can employees browse & enroll themselves?
  requireApproval: boolean // does enrollment need admin approval?
  customBranding: boolean // white-label support
  ssoEnabled: boolean
  brandColors?: {
    primary: string
    secondary: string
  }
}

export interface OrgMembership {
  id: string
  userId: string
  organizationId: string
  role: OrgRole
  department?: string
  joinedAt: Date
  active: boolean
}

/**
 * Links a course to an organization.
 * Controls visibility and assignment within the tenant.
 */
export interface OrgCourse {
  id: string
  organizationId: string
  courseId: string
  assignedTo: 'all' | 'specific' // available to all employees or specific ones
  assignedUserIds?: string[] // only if assignedTo === 'specific'
  assignedDepartments?: string[] // alternative: assign by department
  mandatory: boolean // required training vs optional
  dueDate?: Date // deadline for mandatory courses
  addedBy: string // admin userId who added it
  addedAt: Date
}

/**
 * Tracks per-org analytics and usage
 */
export interface OrgAnalytics {
  organizationId: string
  totalEmployees: number
  activelearners: number
  coursesAssigned: number
  averageCompletion: number
  topCourses: { courseId: string; completionRate: number }[]
}

File: backend/app/models/organization.py
from sqlalchemy import Column, String, Integer, Boolean, DateTime, JSON, Enum as SQLEnum, ForeignKey
from sqlalchemy.orm import relationship
from datetime import datetime
import enum
from app.core.database import Base


class OrgPlan(str, enum.Enum):
    STARTER = "starter"
    PROFESSIONAL = "professional"
    ENTERPRISE = "enterprise"


class OrgRole(str, enum.Enum):
    OWNER = "owner"
    ADMIN = "admin"
    INSTRUCTOR = "instructor"
    LEARNER = "learner"


class Organization(Base):
    __tablename__ = "organizations"

    id = Column(String, primary_key=True, index=True)
    name = Column(String, nullable=False)
    slug = Column(String, unique=True, nullable=False, index=True)
    logo = Column(String, nullable=True)
    domain = Column(String, nullable=True)
    plan = Column(SQLEnum(OrgPlan), default=OrgPlan.STARTER)
    settings = Column(JSON, default={})
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    memberships = relationship("OrgMembership", back_populates="organization", cascade="all, delete-orphan")
    courses = relationship("OrgCourse", back_populates="organization", cascade="all, delete-orphan")


class OrgMembership(Base):
    __tablename__ = "org_memberships"

    id = Column(String, primary_key=True, index=True)
    user_id = Column(String, nullable=False, index=True)
    organization_id = Column(String, ForeignKey("organizations.id"), nullable=False, index=True)
    role = Column(SQLEnum(OrgRole), default=OrgRole.LEARNER)
    department = Column(String, nullable=True)
    active = Column(Boolean, default=True)
    joined_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    organization = relationship("Organization", back_populates="memberships")


class OrgCourse(Base):
    __tablename__ = "org_courses"

    id = Column(String, primary_key=True, index=True)
    organization_id = Column(String, ForeignKey("organizations.id"), nullable=False, index=True)
    course_id = Column(String, nullable=False, index=True)
    assigned_to = Column(String, default="all")  # 'all' or 'specific'
    assigned_user_ids = Column(JSON, default=[])
    assigned_departments = Column(JSON, default=[])
    mandatory = Column(Boolean, default=False)
    due_date = Column(DateTime, nullable=True)
    added_by = Column(String, nullable=False)
    added_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    organization = relationship("Organization", back_populates="courses")
